tools/chapel-py/chplcheck/driver.py

```diff
@@ -0,0 +1,99 @@
+#
+# Copyright 2020-2023 Hewlett Packard Enterprise Development LP
+# Copyright 2004-2019 Cray Inc.
+# Other additional copyright holders may be indicated within.
+#
+# The entirety of this work is licensed under the Apache License,
+# Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License.
+#
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import chapel
+import chapel.core
+
+IgnoreAttr = ("chplcheck.ignore", ["rule", "comment"])
+def ignores_rule(node, rulename):
+    ag = node.attribute_group()
+
+    if ag is None: return False
+    for attr in ag:
+        attr_call = chapel.parse_attribute(attr, IgnoreAttr)
+        if attr_call is None: continue
+
+        ignored_rule = attr_call["rule"]
+        if ignored_rule is not None and ignored_rule.value() == rulename:
+            return True
+
+    return False
+
+class LintDriver:
+    def __init__(self):
+        self.SilencedRules = []
+        self.BasicRules = []
+        self.AdvancedRules = []
+
+    def silence_rules(self, *rules):
+        self.SilencedRules.extend(rules)
+
+    def should_check_rule(self, node, rulename):
+        if rulename in self.SilencedRules:
+            return False
+
+        if node is not None and ignores_rule(node, rulename):
+            return False
+
+        return True
+
+    def check_basic_rule(self, context, root, rule):
+        (name, nodetype, func) = rule
+
+        if not self.should_check_rule(None, name):
+            return
+
+        # If we should ignore the rule no matter the node, no reason to run
+        # a traversal and match the pattern.
```

Comment:
Based on the method below, this comment should be attached to the above if block.

Considering the two check methods have this same early opt out, maybe it should be its own method that gets called? If I understand right, you are checking for if it is silenced? So maybe an alias for `self.should_check_rule(None, name)` is `self.rule_is_silenced(name)`?

Reply:
There's so little duplication there, and just enough dynamic typing (`rule` is different each time, so unpacking it shouldn't be done), that I think factoring the early return is not worth it.

Reply:
I've moved the comment and made `should_check_rule` have a default value (`None`) for node to clean up the shared code in the two check methods.

Reply:
I think that is an excellent solution